### PR TITLE
WIP: client-go: generation based mutation cache

### PR DESCRIPTION
**Pull request by stevekuznetsov**
client-go: expose a generic comparison interface

The ResourceVersion of an object is meant to be entirely opaque,
allowing only equality checks on the client side. The mutation cache
should therefore expose a comparison interface that does not expect to
be using forbidden mechanisms for comparison.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

client-go: allow using generations in mutation cache

If a controller is setting the status.observedGeneration to indicate how
recently it has reacted to the spec of an object, the mutation cache can
track mutations almost as well as with the resourceVersion. In all cases
where the observedGeneration in the mutation is newer than the
generation from the lister, we know that the mutation is newer. When the
two are identical, we cannot tell if the objects are at identical
versions or if the status has been updated without bumping the observed
generation, but we know at the very least that we can use the object
from the lister as it is at least as new as that in the mutation. If the
generation in the lister is newer, we know that the lister's object is
newer.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

```release-note
NONE
```

/kind feature
/cc @sttts

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Open question: do we need to handle observed generations in conditions?

File: staging/src/k8s.io/client-go/tools/cache/mutation_cache.go [hunk 1]
```diff
@@ -42,11 +42,20 @@ type MutationCache interface {
 	Mutation(interface{})
 }
 
-// ResourceVersionComparator is able to compare object versions.
+// ResourceVersionComparator is deprecated.
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/mutation_cache.go:45]: Compare https://go.dev/blog/godoc for how to label this as deprecated.

```
Deprecated: resource version are opaque by definition. Use generations or other fields instead.
```

Comment (sttts) [reply staging/src/k8s.io/client-go/tools/cache/mutation_cache.go:45]: Also deprecate NewIntegerResourceVersionMutationCache

File: staging/src/k8s.io/client-go/tools/cache/mutation_cache.go [hunk 2]
```diff
@@ -260,3 +289,67 @@ func (a etcdObjectVersioner) CompareResourceVersion(lhs, rhs runtime.Object) int
...
+	accessor, ok := obj.(ObservedGenerationAccesor)
+	if !ok {
+		return 0, fmt.Errorf("%T does not expose observed generations", obj)
+	}
+	var latestObservedGeneration int64
+	for _, observedGeneration := range accessor.ObservedGenerations() {
+		if observedGeneration > latestObservedGeneration {
+			latestObservedGeneration = observedGeneration
+		}
+	}
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/mutation_cache.go:320]: trying to reason about whether this is the right behaviour to take the maximum. If there are multiple controllers, each responsible for a subset of observed generations, I think each want to compare with "its own" observed generations.

Imagine controller A and B, writing `status.observedGeneration{A,B}`. `ObjectMeta.generation` changes to 42, with `status.observedGeneration{A,B} == 41`. Controller A observes 42, updates the object status with `status.observedGenerationA = 42`. Controller B observes 42, but finds that `observedGeneration(obj) == 42` already. So B ignores the generation 42.

Comment (stevekuznetsov) [reply staging/src/k8s.io/client-go/tools/cache/mutation_cache.go:320]: Good point, each controller with their own cache will need a controller-specific getter for the observed generation, which makes me think that even more that a per-use implementation of the getter is what we want. I'll refactor, thanks.

**Comment by dims** [comment]
please reopen if you still need this WIP PR. Trying to declutter the open PR(s)

**Pull request closed without merging**